crates/turborepo-cache/src/cache_archive/restore_symlink.rs

```diff
@@ -0,0 +1,150 @@
+
+    Ok(processed_name)
+}
+
+fn actually_restore_symlink<'a>(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    processed_name: &'a AnchoredSystemPath,
+    header: &tar::Header,
+) -> Result<&'a AnchoredSystemPath, CacheError> {
+    dir_cache.safe_mkdir_file(anchor, processed_name)?;
+
+    let symlink_from = anchor.resolve(processed_name);
+
+    _ = symlink_from.remove();
+
+    let link_name = header.link_name()?.expect("have linkname");
+    let symlink_to = link_name.to_str().ok_or_else(|| {
+        CacheError::PathError(
+            PathError::InvalidUnicode(link_name.to_string_lossy().to_string()),
+            Backtrace::capture(),
+        )
+    })?;
+
+    if Utf8Path::new(symlink_to).is_dir() {
+        symlink_from.symlink_to_file(symlink_to)?;
+    } else {
+        symlink_from.symlink_to_dir(symlink_to)?;
+    }
+
+    #[cfg(target_os = "macos")]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        let metadata = symlink_from.symlink_metadata()?;
+        let mut permissions = metadata.permissions();
+        if let Ok(mode) = header.mode() {
+            permissions.set_mode(mode);
+        }
+    }
+
+    Ok(processed_name)
+}
+
+// canonicalizeLinkname determines (lexically) what the resolved path on the
+// system will be when linkname is restored verbatim.
+pub fn canonicalize_linkname(
+    anchor: &AbsoluteSystemPath,
+    processed_name: &AnchoredSystemPathBuf,
+    linkname: &Path,
+) -> Result<Utf8PathBuf, CacheError> {
+    // We don't know _anything_ about linkname. It could be any of:
+    //
+    // - Absolute Unix Path
+    // - Absolute Windows Path
+    // - Relative Unix Path
+    // - Relative Windows Path
+    //
+    // We also can't _truly_ distinguish if the path is Unix or Windows.
+    // Take for example: `/Users/turbobot/weird-filenames/\foo\/lol`
+    // It is a valid file on Unix, but if we do slash conversion it breaks.
+    // Or `i\am\a\normal\unix\file\but\super\nested\on\windows`.
+    //
+    // We also can't safely assume that paths in link targets on one platform
+    // should be treated as targets for that platform. The author may be
+    // generating an artifact that should work on Windows on a Unix device.
+    //
+    // Given all of that, our best option is to restore link targets _verbatim_.
+    // No modification, no slash conversion.
+    //
+    // In order to DAG sort them, however, we do need to canonicalize them.
+    // We canonicalize them as if we're restoring them verbatim.
+    //
+    let cleaned_linkname = Utf8PathBuf::try_from(clean(linkname)).map_err(|_| {
+        CacheError::PathError(
+            PathError::InvalidUnicode(linkname.to_string_lossy().to_string()),
+            Backtrace::capture(),
+        )
+    })?;
+
+    // 1. Check to see if the link target is absolute _on the current platform_.
+    // If it is an absolute path it's canonical by rule.
+    if cleaned_linkname.is_absolute() {
```

Comment:
How about `turbopath::categorize(Utf8Path) -> Either<AnchoredSystemPathBuf, AbsoluteSystemPathBuf>`?

Then we aren't operating directly on `Utf8Path` instances (which we should likely treat like `Path` and try to avoid outside the path library), and we don't need to expose `AnchoredSystemPathBuf::new_unchecked`?